src/configurationProvider.ts

```diff
@@ -213,8 +212,90 @@ export class JavaDebugConfigurationProvider implements vscode.DebugConfiguration
         return Object.keys(config).filter((key: string) => key !== "noDebug").length === 0;
     }
 
-    private async chooseMainClass(folder: vscode.WorkspaceFolder | undefined): Promise<IMainClassOption | undefined> {
-        const res = await resolveMainClass(folder ? folder.uri : undefined);
+    private isEmptyArray(configItems: any): boolean {
+        return !Array.isArray(configItems) || !configItems.length;
+    }
+
+    private async resolveLaunchConfig(folder: vscode.Uri | undefined, config: vscode.DebugConfiguration): Promise<IMainClassOption> {
```

Comment:
If the first parameter is optional, then we could write it as `folder?: vscode.Uri`. The question mark after the param name will assign `undefined` value when the param is not omitted. "default to undefined" is not necessariy by the nature of ts/js.

I suggest that you place `config` param as the first param because it's required. And folder to the second because it's optional.

Reply:
Actually the pattern is inherited from VSCode API. See the original VSCode API defintion `public resolveDebugConfiguration(folder: vscode.WorkspaceFolder | undefined, config: vscode.DebugConfiguration, token?: vscode.CancellationToken):`

Reply:
It indicates the `folder` parameter is not optional. Just represent it could be `undefined` value when no workspace is opened.